runtime-rs: built-in Dragonball sandbox part III - virtio-blk, virtio-fs, virtio-net and VMM API support

studychao opened the pull request:
In this PR, we introduce the abstractions for the virtio-blk, virtio-fs, virtio-net manager, and VMM API support for the runtime to communicate with VMM in one process.

jodh-intel commented on src/dragonball/src/device_manager/fs_dev_mgr.rs:
```diff
@@ -0,0 +1,517 @@
+    /// Number of thread pool workers.
+    pub thread_pool_size: u16,
+    /// Cache policy
+    pub cache_policy: String,
```
What values can this take? Is there a URL you could specify here for further information?

studychao replied:
comment and URL added:)

jodh-intel commented on src/dragonball/src/device_manager/memory_region_handler.rs:
```diff
@@ -0,0 +1,110 @@
+        region: Arc<GuestRegionMmap>,
+    ) -> std::result::Result<(), VirtIoError> {
+        let file_offset = match region.file_offset() {
+            // TODO: use from_arc
```
?

studychao replied:
This is a trivial code refractor we want to do during cleaning up the code.
I'll create an issue for this.

studychao replied:
#4577

jodh-intel commented on src/dragonball/src/event_manager.rs:
```diff
@@ -0,0 +1,169 @@
+pub(crate) struct EventContext {
+    pub api_event_fd: EventFd,
+    pub api_event_flag: bool,
+    pub exit_evt_flag: bool,
```
I think `exit_vmm` would be clearer (here and in the docs).

studychao replied:
Thanks:) I have adopted @bergwolf 's advice to rename those two flags to `api_event_triggered` and `exit_evt_triggered `.

jodh-intel commented on src/dragonball/src/event_manager.rs:
```diff
@@ -0,0 +1,169 @@
+/// Shared information between vmm::vmm_thread_event_loop() and VmmEpollHandler.
+pub(crate) struct EventContext {
+    pub api_event_fd: EventFd,
+    pub api_event_flag: bool,
```
Would `api_event_fd_invalid` be clearer here? But atually, maybe you could store the error there instead? Something like:

```suggestion
    pub api_event_error: Option<Error>,
```

studychao replied:
This `api_event_flag` is for triggering api event handling in `run_vmm_event_loop` so not only error situation will be handled here. Other requests like `ShutdownMicroVm`, `GetVmConfiguration` will also be handled.

jodh-intel reviewed (commented):
Thanks @studychao - a few comments...

bergwolf commented on src/dragonball/src/io_manager.rs:
```diff
@@ -0,0 +1,60 @@
+    #[cfg(target_arch = "x86_64")]
+    #[inline]
+    /// Read data from IO ports.
+    pub fn pio_read(&mut self, addr: u16, data: &mut [u8]) -> Result<()> {
```
What are IO ports? Why does it only work for x86_64?

studychao replied:
You can find the introduction of io ports at https://wiki.osdev.org/I/O_Ports. This is an x86-specific feature and aarch64 uses mmio to deal with device data transactions.

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,670 @@
+/// There are two level of API servers as below:
+/// API client <--> VMM API Server <--> VMM Core
+pub struct VmmService {
+    _api_event_fd: EventFd,
```
Because `_api_event_fd` is not used in current code. I'll remove this.:)

studychao commented on src/dragonball/src/vcpu/vcpu_impl.rs:
```diff
@@ -0,0 +1,959 @@
+
+    #[cfg(target_arch = "aarch64")]
+    /// Error getting the Vcpu preferred target on Arm.
+    #[error("failure while getting the vCPU preferred target on Arm")]
```
fixed in PART II PR:)

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,670 @@
+    fn send_response(&self, result: VmmRequestResult) -> Result<()> {
+        self.to_api
+            .send(Box::new(result))
+            .map_err(|_| ())
+            .expect("vmm: one-shot API result channel has been closed");
```
Because we think vmm event handle thread disconnection is a serious problem and we will panic the vmm event handle thread

studychao commented on src/dragonball/src/api/v1/boot_source.rs:
```diff
@@ -0,0 +1,53 @@
+/// Strongly typed data structure used to configure the boot source of the microvm.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize, Default)]
+#[serde(deny_unknown_fields)]
+pub struct BootSourceConfig {
```
I'll work on this, might take some time~

studychao commented on src/dragonball/src/api/v1/boot_source.rs:
```diff
@@ -0,0 +1,53 @@
+    /// Path of the kernel image.
+    pub kernel_image_path: String,
+    /// Path of the initrd, if there is one.
+    pub initrd_path: Option<String>,
```
Comment for rootfs added.

studychao commented on src/dragonball/src/api/v1/boot_source.rs:
```diff
@@ -0,0 +1,53 @@
+#[serde(deny_unknown_fields)]
+pub struct BootSourceConfig {
+    /// Path of the kernel image.
+    pub kernel_image_path: String,
```
1. Changed to `kernel_path`
2. kernel format explanation added.

studychao commented on src/dragonball/src/device_manager/fs_dev_mgr.rs:
```diff
@@ -0,0 +1,517 @@
+
+    /// attach backend fs must be done when vm is running.
+    #[error("vm is not running when attaching a backend fs")]
+    MicroVMMustRunning,
```
Changed.

studychao commented on src/dragonball/src/device_manager/fs_dev_mgr.rs:
```diff
@@ -0,0 +1,517 @@
+            tag: String::default(),
+            num_queues: 1,
+            queue_size: 1024,
+            cache_size: 2 * 1024 * 1024 * 1024,
```
Const created for this

studychao commented on src/dragonball/src/event_manager.rs:
```diff
@@ -0,0 +1,169 @@
+pub(crate) struct EventContext {
+    pub api_event_fd: EventFd,
+    pub api_event_flag: bool,
+    pub exit_evt_flag: bool,
```
Thanks, renamed.

studychao commented:
Hi, @jodh-intel @bergwolf I have addressed all your comments except `safe_path` one.
I'll take some time to use `safe_path` lib.
Please take a look at this PR again when you have some time.

jodh-intel commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,769 @@
+            ));
+        }
+
+        let mut index = 0;
+        // check whether the Device Config belongs to a root device
+        // we need to satisfy the condition by which a VMM can only have on root device
+        if block_device_config.is_root_device {
+            if self.has_root_block {
+                return Err(BlockDeviceError::RootBlockDeviceAlreadyAdded);
+            } else {
+                self.has_root_block = true;
+                self.read_only_root = block_device_config.is_read_only;
+                self.has_partuuid_root = block_device_config.partuuid.is_some();
+                self.partuuid = block_device_config.partuuid.clone();
+                // Root Device should be the first in the list whether or not PARTUUID is specified
+                // in order to avoid bugs in case of switching from partuuid boot scenarios to
+                // /dev/vda boot type.
+                self.info_list
+                    .push_front(BlockDeviceInfo::new(block_device_config));
+            }
+        } else {
+            self.info_list
+                .push_back(BlockDeviceInfo::new(block_device_config));
+            index = self.info_list.len() - 1;
+        }
+
+        Ok(index)
```
This code looks a bit odd to me: why not just do this?:

```diff
diff --git a/src/dragonball/src/device_manager/blk_dev_mgr.rs b/src/dragonball/src/device_manager/blk_dev_mgr.rs
index 3afab9c63..b47e3c2df 100644
--- a/src/dragonball/src/device_manager/blk_dev_mgr.rs
+++ b/src/dragonball/src/device_manager/blk_dev_mgr.rs
@@ -612,7 +612,6 @@ impl BlockDeviceMgr {
             ));
         }

-        let mut index = 0;
         // check whether the Device Config belongs to a root device
         // we need to satisfy the condition by which a VMM can only have on root device
         if block_device_config.is_root_device {
@@ -628,14 +627,13 @@ impl BlockDeviceMgr {
                 // /dev/vda boot type.
                 self.info_list
                     .push_front(BlockDeviceInfo::new(block_device_config));
+                Ok(0)
             }
         } else {
             self.info_list
                 .push_back(BlockDeviceInfo::new(block_device_config));
-            index = self.info_list.len() - 1;
+            Ok(self.info_list.len() - 1)
         }
-
-        Ok(index)
     }

     /// Updates a Block Device Config. The update fails if it would result in two
```

studychao replied:
That looks clearer!!
Thanks, changed.

jodh-intel commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,769 @@
+            }
+        }
+        // Update the config.
+        self.info_list[index].config = new_config;
```
Call it paranoia but I personally don't think we should ever index directly like this: it's too fragile if the surrounding code changes (correct today, but who knows about tomorrow...?)

tbh I'm kinda amazed rust even allows it given it's safety assurances as this just side-steps all that and crashes on invalid indices...

studychao replied:
index check added here.

Yes, I am also amazed that Rust doesn't have a safety check for this scenario. :o

jodh-intel commented on src/dragonball/src/device_manager/fs_dev_mgr.rs:
```diff
@@ -0,0 +1,523 @@
+    pub drop_sys_resource: bool,
+    /// virtio fs or vhostuser fs.
+    pub mode: String,
+    /// Enable kill_priv_v2 or not
+    pub killpriv_v2: bool,
```
- What is this?
- Why v2? (what was v1?)
- How is this kill privilege granted?
- What can be killed if this is `true`?

studychao replied:
This is a fuse feature flag and you could get more information here: https://patchwork.kernel.org/project/linux-fsdevel/patch/20200916161737.38028-2-vgoyal@redhat.com/

jodh-intel commented on src/dragonball/src/vmm.rs:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    /// Get a reference to a virtual machine managed by the VMM.
+    pub fn get_vm_by_id(&self, _id: &str) -> Option<&Vm> {
+        Some(&self.vm)
+    }
+
+    /// Get a mutable reference to a virtual machine managed by the VMM.
+    pub fn get_vm_by_id_mut(&mut self, _id: &str) -> Option<&mut Vm> {
```
Why are these methods used? There are lots of calls to and they all specify an empty id (`get_vm_by_id_mut("")`), so why is the id param needed? The current calls look odd so why not just call these methods `get_vm()` and `get_vm_mut()`?

studychao replied:
Changed.

jodh-intel reviewed (commented):
Thanks @studychao - a few more comments...

jodh-intel reviewed (commented):
Thanks @studychao - a few more comments...

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,664 @@
+        config.mem_type = machine_config.mem_type;
+
+        let mem_size_mib_value = machine_config.mem_size_mib;
+        // Support 1TB memory at most, 2MB aligned for huge page.
```
I have added document in api.md :)

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,664 @@
+
+    /// The action `StartMicroVm` failed either because of bad user input or an internal error.
+    #[error("failed to boot the vm: {0}")]
+    StartMicrovm(#[source] StartMicrovmError),
```
Thanks, changed.

studychao commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,769 @@
+    pub is_root_device: bool,
+    /// Part-UUID. Represents the unique id of the boot partition of this device.
+    /// It is optional and it will be used only if the `is_root_device` field is true.
+    pub partuuid: Option<String>,
```
Thanks, changed.

studychao commented on src/dragonball/src/vmm.rs:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    /// Get a mutable reference to a virtual machine managed by the VMM.
+    pub fn get_vm_by_id_mut(&mut self, _id: &str) -> Option<&mut Vm> {
```
Changed.

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,664 @@
+            .map_err(StartMicrovm)
+    }
+
+    fn shutdown_microvm(&mut self, vmm: &mut Vmm) -> VmmRequestResult {
```
Currently, vmm_event_loop will detect `exit_evt_triggered` and shutdown Dragonball.
If adding tracing needs blocking here, we could consider to add some mechanism here.

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,664 @@
+        let vm_id = vm.shared_info().read().unwrap().id.clone();
+        let serial_path = match machine_config.serial_path {
+            Some(value) => value,
+            None => {
+                if config.serial_path.is_none() {
+                    String::from("/tmp/") + &vm_id + "_com1"
+                } else {
+                    // Safe to unwrap() because we have checked it has a value.
+                    config.serial_path.as_ref().unwrap().clone()
+                }
+            }
```
why is this being automatically set -》 We want to create the console for users even if they didn't set a socket path so that users could enter the guest.

studychao replied:
I have changed the path to "/run/dragonball/..."

studychao commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,769 @@
+#[derive(Debug, thiserror::Error)]
+pub enum BlockDeviceError {
+    /// Invalid virtual machine instance ID.
+    #[error("invalid virtual machine instance id")]
```
Changed

studychao commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,769 @@
+impl BlockDeviceType {
+    /// Get type of low level storage device/protocol by parsing `path`.
+    pub fn get_type(path: &str) -> BlockDeviceType {
+        if path.starts_with("spool") {
```
You could learn more about SPOOL here : https://www.usenix.org/conference/atc20/presentation/xue .
Basically, it's a reliable NVMe virtualization system for the cloud environment.

studychao replied:
I'll add comment here.

studychao commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,769 @@
+                    .map_err(DeviceMgrError::Cmdline)?;
+            }
+            if self.read_only_root {
+                cmdline.insert_str("ro").map_err(DeviceMgrError::Cmdline)?;
```
rw check added and error will be returned.

studychao commented:
Hi @jodh-intel , thanks for your review. I have addressed all your comments, please take a look at it.

jodh-intel commented on src/dragonball/src/device_manager/fs_dev_mgr.rs:
```diff
@@ -0,0 +1,524 @@
+    fn check_conflicts(&self, other: &Self) -> Result<(), FsDeviceError> {
+        if self.tag == other.tag {
+            Err(FsDeviceError::FsDeviceTagAlreadyExists(self.tag.clone()))
+        } else if self.mode.as_str() == "vhostuser" && self.sock_path == other.sock_path {
```
Nit: Might be worth making `"vhostuser"` a top-level `const` with a comment explaining where this string value comes from for clarity.

studychao replied:
const added for vhostuser and virtio.

jodh-intel commented on src/dragonball/src/device_manager/mod.rs:
```diff
@@ -515,16 +598,66 @@ impl DeviceManager {
+    }
 
+    /// Start all registered devices when booting the associated virtual machine.
+    pub fn start_devices(&mut self) -> std::result::Result<(), StartMicroVmError> {
```
When will this method be implemented? Is there another open PR for this?

studychao replied:
This will be implemented in the future stages of vfio support.
I'll create an issue and a TODO here.

jodh-intel commented on src/dragonball/src/vm/mod.rs:
```diff
@@ -89,8 +144,659 @@ impl Default for VmConfigInfo {
+        Ok(())
+    }
+
+    /// Resume all vcpus and calc the intance downtime
```
- Nit: typo: "instance"
- The start of the downtime timestamp is recorded in `pause_all_vcpus_with_downtime()` but I don't see where the overall downtime calculation is...?

studychao replied:
Added.

jodh-intel commented on src/dragonball/src/vm/mod.rs:
```diff
@@ -89,8 +144,659 @@ impl Default for VmConfigInfo {
+            return Ok(());
+        }
+
+        // vcpu boot up require local memory. reserve 100 MiB memory
```
ooi, is this value generic or does it depend on architecture?

studychao replied:
We found that trait in both x86_64 and aarch64 .

jodh-intel commented on src/dragonball/src/vm/mod.rs:
```diff
@@ -89,8 +144,659 @@ impl Default for VmConfigInfo {
+        if mem_type == "hugetlbfs" {
+            let shared_info = self.shared_info.read()
+                .expect("Failed to determine if instance is initialized because shared info couldn't be read due to poisoned lock");
+            mem_file_path.push('/');
+            mem_file_path.push_str(shared_info.id.as_str());
```
Should this path include something like `dragonball`?

studychao replied:
Changed.

jodh-intel commented on src/dragonball/src/vm/x86_64.rs:
```diff
@@ -0,0 +1,279 @@
+            .set_reset_event_fd(reset_event_fd)
+            .map_err(StartMicroVmError::Vcpu)?;
+
+        if self.vm_config.cpu_pm == "on" {
+            info!(self.logger, "VM: enable CPU disable_idle_exits capability");
+        }
```
- What values can `cpu_pm` take? It looks rather odd to be checking a string value `"on"`.
- This doesn't seem to do anything apart from make a log call :) Will code be added in a follow-up PR?

studychao replied:
#4590

jodh-intel commented on src/dragonball/src/vmm.rs:
```diff
@@ -0,0 +1,215 @@
+        .unwrap()
+    }
+
+    #[test]
+    fn test_create_vmm_instance() {
```
This is a **huge** PR but this is the only new test fwics...? Please could you consider writing more unit tests to help us have confidence in this new code? I'd also suggest breaking up a lot of the functions which are imho too big.

studychao replied:
In order to reduce the complexity and size of this PR, we have stripped the unit tests from this PR. We'll create a PR (which is the next Dragonball PR) for the unit tests.

studychao replied:
still need some time to rebase the code:( will do that tomorrow.

studychao replied:
#4604

jodh-intel reviewed (commented):
Thanks @studychao. A few more comments...

studychao commented on src/dragonball/src/api/v1/machine_config.rs:
```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/// We only support 254 vcpus for now. Mostly because we have set all vcpu related metrics as u8
```
Thanks, changed.

studychao commented on src/dragonball/src/device_manager/fs_dev_mgr.rs:
```diff
@@ -0,0 +1,524 @@
+    pub tag: String,
+    /// Number of virtqueues to use.
+    pub num_queues: usize,
+    /// Size of each virtqueue.
```
Comment added.

studychao commented on src/dragonball/src/device_manager/blk_dev_mgr.rs:
```diff
@@ -0,0 +1,775 @@
+                    } else {
+                        info!(
+                            ctx.logger(),
+                            "Queue {}: Using Aio Raw disk file, queue size {}", i, queue_size
```
Thanks, added.

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,629 @@
+/// Wrapper for all errors associated with VMM actions.
+#[derive(Debug, thiserror::Error)]
+pub enum VmmActionError {
+    /// Invalid virutal machine instance ID.
```
Changed.

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,629 @@
+
+/// VMM Service to handle requests from the API server.
+///
+/// There are two level of API servers as below:
```
Changed.

studychao commented on src/dragonball/src/api/v1/machine_config.rs:
```diff
@@ -0,0 +1,84 @@
+    UpdateNotAllowedPostBoot,
+
+    /// The max vcpu count is invalid.
+    #[error("the vCPU number shouldn't large than 254")]
```
MAX_SUPPORTED_VCPUS added in the error log.

studychao commented on src/dragonball/src/api/v1/machine_config.rs:
```diff
@@ -0,0 +1,84 @@
+    InvalidMemorySize(usize),
+
+    /// The hotplug memory size is invalid. The memory can only be an unsigned integer.
+    #[error("the hotplug memory size (MiB) is invalid, must be multiple of 64MB")]
```
const added in the error log

studychao commented on src/dragonball/src/api/v1/vmm_action.rs:
```diff
@@ -0,0 +1,629 @@
+            config.vcpu_count = vcpu_count;
+        }
+
+        if config.cpu_topology != machine_config.cpu_topology {
```
Thanks, added.

studychao commented on src/dragonball/src/device_manager/memory_region_handler.rs:
```diff
@@ -0,0 +1,110 @@
+            VirtIoError::IOError(io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "invalid address space region (0x{0:x}, 0x{1:x})",
```
Thanks, changed, also learn something new :)

studychao commented:
Hi @jodh-intel , I have addressed all your comments, please take a look at them:)

Besides, I'll release the PR for unit tests.

studychao commented:
PR #4604 is created for unit tests:)

jodh-intel reviewed (approved):
Thanks @studychao.

lgtm

bergwolf reviewed (approved):
Thanks!
